docs/topics/concurrency.rst

```diff
@@ -70,40 +75,37 @@ Here is the program in examples/thread_pool_executor.py.
         reversed.
         """
 
-        with rasterio.Env():
-
-            with rasterio.open(infile) as src:
-
-                # Create a destination dataset based on source params. The
-                # destination will be tiled, and we'll process the tiles
-                # concurrently.
-                profile = src.profile
-                profile.update(blockxsize=128, blockysize=128, tiled=True)
-
-                with rasterio.open(outfile, "w", **profile) as dst:
-
-                    # Materialize a list of destination block windows
-                    # that we will use in several statements below.
-                    windows = [window for ij, window in dst.block_windows()]
-
-                    # This generator comprehension gives us raster data
-                    # arrays for each window. Later we will zip a mapping
-                    # of it with the windows list to get (window, result)
-                    # pairs.
-                    data_gen = (src.read(window=window) for window in windows)
-
-                    with concurrent.futures.ThreadPoolExecutor(
-                        max_workers=num_workers
-                    ) as executor:
-
-                        # We map the compute() function over the raster
-                        # data generator, zip the resulting iterator with
-                        # the windows list, and as pairs come back we
-                        # write data to the destination dataset.
-                        for window, result in zip(
-                            windows, executor.map(compute, data_gen)
-                        ):
-                            dst.write(result, window=window)
+        with rasterio.open(infile) as src:
+
+            # Create a destination dataset based on source params. The
+            # destination will be tiled, and we'll process the tiles
+            # concurrently.
+            profile = src.profile
+            profile.update(blockxsize=128, blockysize=128, tiled=True)
+
+            with rasterio.open(outfile, "w", **src.profile) as dst:
+                windows = [window for ij, window in dst.block_windows()]
+
+                # We cannot write to the same file from multiple threads
+                # without causing race conditions. To safely read/write
+                # from multiple threads, we use a lock to protect the
+                # DatasetReader/Writer
+                read_lock = multiprocessing.Lock()
+                write_lock = multiprocessing.Lock()
+
+                def process(window):
+                    with read_lock:
+                        src_array = src.read(window=window)
+                    result = compute(src_array)
```

Comment:
Add some white space around line 99?

I think the important thing to get across is that if `compute()` releases the GIL, it can truly execute concurrently in different worker threads. If `compute()` uses, for example, numpy operations (see https://scipy-cookbook.readthedocs.io/items/ParallelProgramming.html#Threads) it can execute partially concurrently  in different threads.

Reply:
That's true, but even if Threads do not release the GIL, you can get a decent speedup, because 1 thread can be reading, 1 thread computing and 1 thread writing. But this of course depends largely on where the bottleneck is.

I'll read it through one more time to see if I can add some more info on the GIL. I find it quite hard to find information on numpy and multithreading. It seems that in some cases numpy already uses multithreading internally and multithreading it manually is not really needed.

Reply:
That's a good point in your first paragraph.